File: src/core/extractor.ts
import { generateFactId, generateDecisionId, generateActionId, generateEntityId } from './ids.js'
import type { ContextLaneChunk, ContextLaneCitation, ContextLaneExtraction, ContextLaneFact, ContextLaneDecision, ContextLaneAction, ContextLaneEntity } from './schema.js'

export interface ExtractOptions {
  runId: string
  chunks: ContextLaneChunk[]
  citations: ContextLaneCitation[]
}

export function extract(options: ExtractOptions): ContextLaneExtraction {
  const { runId, chunks, citations } = options

  const allText = chunks.map(c => c.text).join('\n')
  const lines = allText.split('\n').filter(l => l.trim())

  const summary = generateSummary(chunks)
  const facts = extractFacts(chunks, citations)
  const decisions = extractDecisions(chunks, citations)
  const actions = extractActions(chunks, citations)
  const entities = extractEntities(chunks, citations)
  const tags = extractTags(chunks)

  return {
    runId,
    summary,
    facts,
    decisions,
    actions,
    entities,
    tags,
    citations,
    createdAt: new Date().toISOString(),
  }
}

function generateSummary(chunks: ContextLaneChunk[]): string {
  const headlines = chunks
    .map(c => c.text.split('\n').filter(l => l.trim().startsWith('#')))
    .flat()
    .slice(0, 10)

  if (headlines.length > 0) {
    return `Source contains ${chunks.length} chunks across ${headlines.length} sections. Top sections: ${headlines.map(h => h.trim().replace(/^#+\s*/, '')).join(', ')}.`
  }

  const firstLines = chunks[0]?.text.split('\n').slice(0, 3).join(' ').slice(0, 200) || ''
  return `Source contains ${chunks.length} chunks. First content: ${firstLines}...`
}

function extractFacts(chunks: ContextLaneChunk[], citations: ContextLaneCitation[]): ContextLaneFact[] {
  const facts: ContextLaneFact[] = []
  const seen = new Set<string>()

  for (const chunk of chunks) {
    const lines = chunk.text.split('\n')
    for (const line of lines) {
      const trimmed = line.trim()
      if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith('```')) continue
      if (trimmed.length < 20) continue

      const key = trimmed.slice(0, 60)
      if (seen.has(key)) continue
      seen.add(key)

      const cit = citations.find(c => c.startLine === chunk.startLine) || citations.find(c => c.sourceId === chunk.sourceId)
      facts.push({
        id: generateFactId(),
        text: trimmed.slice(0, 300),
        confidence: 0.7,
        tags: [],
        citationIds: cit ? [cit.id] : [],
      })
    }
  }

  return facts.slice(0, 50)
}

const ACTION_KEYWORDS = ['TODO', 'FIXME', 'HACK', 'need to', 'should', 'must', 'required', 'action item', 'action', 'todo:', 'to do']
const DECISION_KEYWORDS = ['decision', 'we decided', 'we chose', 'agreed', 'elected', 'resolved', 'conclusion', 'outcome']

function extractDecisions(chunks: ContextLaneChunk[], citations: ContextLaneCitation[]): ContextLaneDecision[] {
  const decisions: ContextLaneDecision[] = []

  for (const chunk of chunks) {
    for (const line of chunk.text.split('\n')) {
      const lower = line.toLowerCase()
      if (DECISION_KEYWORDS.some(k => lower.includes(k))) {
        const cit = citations.find(c => c.startLine === chunk.startLine)
        decisions.push({
          id: generateDecisionId(),
          text: line.trim().slice(0, 300),
          status: 'active',
          citationIds: cit ? [cit.id] : [],
        })
      }
    }
  }

  return decisions
}

function extractActions(chunks: ContextLaneChunk[], citations: ContextLaneCitation[]): ContextLaneAction[] {
  const actions: ContextLaneAction[] = []

  for (const chunk of chunks) {
    for (const line of chunk.text.split('\n')) {
      const lower = line.toLowerCase()
      const matched = ACTION_KEYWORDS.find(k => lower.includes(k))
      if (matched) {
        const priority = matched === 'HACK' || matched === 'FIXME' ? 'high' as const : matched === 'TODO' ? 'medium' as const : 'low' as const
        const cit = citations.find(c => c.startLine === chunk.startLine)
        actions.push({
          id: generateActionId(),
          text: line.trim().slice(0, 300),
          priority,
          citationIds: cit ? [cit.id] : [],
        })
      }
    }
  }

  return actions
}

const URL_RE = /https?:\/\/[^\s)"'\]>]+/g
const PROJECT_RE = /@[\w-]+\/[\w-]+/g
const TOOL_RE = /[\w-]+/g

function extractEntities(chunks: ContextLaneChunk[], citations: ContextLaneCitation[]): ContextLaneEntity[] {
  const entityMap = new Map<string, { type: ContextLaneEntity['type']; mentions: number; citationIds: Set<string> }>()

  for (const chunk of chunks) {
    const matchingCitations = citations.filter(c => c.startLine === chunk.startLine)

    for (const match of chunk.text.matchAll(URL_RE)) {
      const url = match[0]
      const e = entityMap.get(url) || { type: 'url' as const, mentions: 0, citationIds: new Set() }
      e.mentions++
      for (const c of matchingCitations) e.citationIds.add(c.id)
      entityMap.set(url, e)
    }

    for (const match of chunk.text.matchAll(PROJECT_RE)) {
      const name = match[0]
      const e = entityMap.get(name) || { type: 'project' as const, mentions: 0, citationIds: new Set() }
      e.mentions++
      for (const c of matchingCitations) e.citationIds.add(c.id)
      entityMap.set(name, e)
    }
  }

  return Array.from(entityMap.entries()).map(([name, data]) => ({
    id: generateEntityId(),
    name,
    type: data.type,
    mentions: data.mentions,
    citationIds: Array.from(data.citationIds),
  }))
}

function extractTags(chunks: ContextLaneChunk[]): string[] {
  const freq = new Map<string, number>()
  const skip = new Set(['the', 'this', 'that', 'with', 'from', 'have', 'been', 'will', 'what', 'about', 'when', 'where', 'which', 'their', 'there', 'would', 'could', 'should', 'into', 'more', 'some', 'than', 'then', 'also', 'just', 'like', 'very', 'well', 'even', 'still', 'over', 'such', 'only', 'other', 'after', 'before', 'between', 'through', 'during', 'without', 'within', 'along', 'following', 'including', 'using', 'based', 'related', 'various', 'multiple', 'within', 'across', 'among', 'before', 'after', 'above', 'below', 'under', 'while', 'because', 'though', 'unless', 'until', 'since', 'about'])

  for (const chunk of chunks) {
    const words = chunk.text.toLowerCase().split(/\W+/).filter(w => w.length > 3 && !skip.has(w))
    for (const w of words) {
      freq.set(w, (freq.get(w) || 0) + 1)
    }
  }

  return Array.from(freq.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, 15)
    .map(([word]) => word)
}

File: python/contextlane/core/extractor.py
import re
from .ids import generate_fact_id, generate_decision_id, generate_action_id, generate_entity_id


ACTION_KEYWORDS = ["TODO", "FIXME", "HACK", "need to", "should", "must", "required", "action item", "todo:"]
DECISION_KEYWORDS = ["decision", "we decided", "we chose", "agreed", "elected", "resolved", "conclusion", "outcome"]
SKIP_WORDS = {"the", "this", "that", "with", "from", "have", "been", "will", "what", "about",
              "when", "where", "which", "their", "there", "would", "could", "should", "into",
              "more", "some", "than", "then", "also", "just", "like", "very", "well", "even",
              "still", "over", "such", "only", "other", "after", "before", "between", "through",
              "during", "without", "within", "along", "following", "including", "using", "based",
              "related", "various", "multiple", "across", "among", "while", "because", "though",
              "unless", "until", "since"}


def extract(run_id: str, chunks: list, citations: list) -> dict:
    all_text = "\n".join(c["text"] for c in chunks)
    lines = [l for l in all_text.split("\n") if l.strip()]

    summary = _generate_summary(chunks)
    facts = _extract_facts(chunks, citations)
    decisions = _extract_decisions(chunks, citations)
    actions = _extract_actions(chunks, citations)
    entities = _extract_entities(chunks, citations)
    tags = _extract_tags(chunks)

    return {
        "run_id": run_id,
        "summary": summary,
        "facts": facts,
        "decisions": decisions,
        "actions": actions,
        "entities": entities,
        "tags": tags,
        "citations": citations,
        "created_at": __import__("datetime").datetime.now().isoformat(),
    }


def _generate_summary(chunks: list) -> str:
    headlines = []
    for c in chunks:
        for line in c["text"].split("\n"):
            if line.strip().startswith("#"):
                headlines.append(line.strip().lstrip("#").strip())

    if headlines:
        return f"Source contains {len(chunks)} chunks across {len(headlines)} sections: {', '.join(headlines[:10])}."

    first = chunks[0]["text"][:200] if chunks else ""
    return f"Source contains {len(chunks)} chunks. First content: {first}..."


def _extract_facts(chunks: list, citations: list) -> list:
    facts = []
    seen = set()

    for chunk in chunks:
        for line in chunk["text"].split("\n"):
            t = line.strip()
            if not t or t.startswith("#") or t.startswith("```"):
                continue
            if len(t) < 20:
                continue
            key = t[:60]
            if key in seen:
                continue
            seen.add(key)

            cit_ids = [c["id"] for c in citations if c.get("start_line") == chunk.get("start_line")]
            if not cit_ids:
                cit_ids = [c["id"] for c in citations if c["source_id"] == chunk["source_id"]]

            facts.append({
                "id": generate_fact_id(),
                "text": t[:300],
                "confidence": 0.7,
                "tags": [],
                "citation_ids": cit_ids[:1],
            })

    return facts[:50]


def _extract_decisions(chunks: list, citations: list) -> list:
    decisions = []
    for chunk in chunks:
        for line in chunk["text"].split("\n"):
            lower = line.lower()
            if any(k in lower for k in DECISION_KEYWORDS):
                cit_ids = [c["id"] for c in citations if c.get("start_line") == chunk.get("start_line")]
                decisions.append({
                    "id": generate_decision_id(),
                    "text": line.strip()[:300],
                    "status": "active",
                    "citation_ids": cit_ids[:1],
                })
    return decisions


def _extract_actions(chunks: list, citations: list) -> list:
    actions = []
    for chunk in chunks:
        for line in chunk["text"].split("\n"):
            lower = line.lower()
            matched = next((k for k in ACTION_KEYWORDS if k.lower() in lower), None)
            if matched:
                priority = "high" if matched in ("HACK", "FIXME") else "medium" if matched == "TODO" else "low"
                cit_ids = [c["id"] for c in citations if c.get("start_line") == chunk.get("start_line")]
                actions.append({
                    "id": generate_action_id(),
                    "text": line.strip()[:300],
                    "priority": priority,
                    "citation_ids": cit_ids[:1],
                })
    return actions


URL_RE = re.compile(r"https?://[^\s)\"'>\]]+")
PROJECT_RE = re.compile(r"@[\w-]+/[\w-]+")


def _extract_entities(chunks: list, citations: list) -> list:
    entity_map = {}

    for chunk in chunks:
        matching_citations = [c for c in citations if c.get("start_line") == chunk.get("start_line")]
        cit_ids = [c["id"] for c in matching_citations]

        for m in URL_RE.finditer(chunk["text"]):
            url = m.group(0)
            if url not in entity_map:
                entity_map[url] = {"type": "url", "mentions": 0, "citation_ids": set()}
            entity_map[url]["mentions"] += 1
            entity_map[url]["citation_ids"].update(cit_ids)

        for m in PROJECT_RE.finditer(chunk["text"]):
            name = m.group(0)
            if name not in entity_map:
                entity_map[name] = {"type": "project", "mentions": 0, "citation_ids": set()}
            entity_map[name]["mentions"] += 1
            entity_map[name]["citation_ids"].update(cit_ids)

    return [
        {
            "id": generate_entity_id(),
            "name": name,
            "type": data["type"],
            "mentions": data["mentions"],
            "citation_ids": list(data["citation_ids"]),
        }
        for name, data in entity_map.items()
    ]


def _extract_tags(chunks: list) -> list:
    freq = {}
    for chunk in chunks:
        words = re.findall(r"\w{4,}", chunk["text"].lower())
        for w in words:
            if w not in SKIP_WORDS:
                freq[w] = freq.get(w, 0) + 1

    return [w for w, _ in sorted(freq.items(), key=lambda x: -x[1])[:15]]
